lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -1210,6 +1288,408 @@ void BlockedEncodingAttr::print(mlir::AsmPrinter &printer) const {
+  assert(llvm::all_of(
+      llvm::zip(ll.getOutDimNames(), llvm::seq(ll.getNumOutDims())),
+      [](auto pair) {
+        auto [outDim, i] = pair;
+        return outDim.str() == ("dim" + llvm::Twine(i)).str();
+      }));
+
+  printer << "<{bases = {"
+          << join(ll.getBases(), ", ",
+                  [](const auto &base) {
+                    return base.first.str() + " = " + "[" +
+                           join(base.second, ", ",
+                                [](const std::vector<int32_t> &vec) {
+                                  return "[" + join(vec, ", ") + "]";
+                                }) +
+                           "]";
+                  })
+          << "}, repOrder = [" << mlir::triton::join(getRepOrder(), ", ")
+          << "]}>";
+}
+
+Attribute LinearEncodingAttr::parse(AsmParser &parser, Type type) {
+  if (parser.parseLess().failed())
+    return {};
+
+  DictionaryAttr dict;
+  if (parser.parseAttribute(dict).failed())
+    return {};
+
+  if (parser.parseGreater().failed())
+    return {};
+
+  LinearLayout::BasesT bases;
+
+  // Parse the basis names in order (the order is relevant)
+  std::vector<std::string> inDimNames = {"register", "lane", "warp", "block"};
+  auto dictBases = dict.getAs<DictionaryAttr>("bases");
+  if (!dictBases) {
+    parser.emitError(parser.getCurrentLocation(), "Expected bases");
+    return {};
+  }
+
+  for (const auto &inDimNameStr : inDimNames) {
+    auto inDimName = StringAttr::get(parser.getContext(), inDimNameStr);
+    Attribute value = dictBases.get(inDimName);
+
+    // Expecting an array of arrays
+    auto arrayOfArraysAttr = mlir::dyn_cast<ArrayAttr>(value);
+    if (!arrayOfArraysAttr) {
+      parser.emitError(parser.getCurrentLocation(),
+                       "Expected array of arrays for basis of '")
+          << inDimName.getValue() << "'";
+      return {};
+    }
+
+    std::vector<std::vector<int32_t>> inDimBases;
+    for (Attribute arrayAttr : arrayOfArraysAttr) {
+      auto intArrayAttr = mlir::dyn_cast<ArrayAttr>(arrayAttr);
+      if (!intArrayAttr) {
+        parser.emitError(parser.getCurrentLocation(),
+                         "Expected array of integers in basis for '")
+            << inDimName.getValue() << "'";
+        return {};
+      }
+      std::vector<int32_t> basis;
+      for (Attribute intAttr : intArrayAttr) {
+        auto intValueAttr = mlir::dyn_cast<IntegerAttr>(intAttr);
+        if (!intValueAttr) {
+          parser.emitError(parser.getCurrentLocation(),
+                           "Expected integer in basis for '")
+              << inDimName.getValue() << "'";
+          return {};
+        }
+        basis.push_back(intValueAttr.getInt());
+      }
+      inDimBases.push_back(std::move(basis));
+    }
+    bases[inDimName] = std::move(inDimBases);
+  }
+
+  auto lanes = bases.find(StringAttr::get(parser.getContext(), "lane"));
+  assert(!lanes->second.empty() && "There must be at least two lanes!");
```

Comment:
> There must be at least two lanes!

Why there must be two lanes?

Reply:
I need to get the rank of the tensor from somewhere. If an input dim has `size=1`, then the list of bases is empty and I can't infer the rank from it. I figured out that the only input dim that cannot have `size=1` is the lanes, as there are 32 lanes per warp (or 64 in AMD).

A more portable way here would be to assert that the layout is not empty, and then take the largest rank out of all the input dims. Is it fine to leave a comment describing this, or would you want me to implement this?

Reply:
> Is it fine to leave a comment describing this, or would you want me to implement this?

I think I get your point, but what about triton CPU which may have a single thread in the layout?